adls/src/adls_client.cpp

```diff
@@ -0,0 +1,391 @@
+        http->set_is_input_length_known();
+        http->set_input_content_length(stream_len);
+
+        auto async_func = std::bind(&azure::storage_lite::async_executor<void>::submit, m_account, request, http, m_context);
+        return blob_client_adaptor<void>(async_func);
+    }
+
+    void adls_client::flush_data(const std::string& filesystem, const std::string& file, uint64_t offset)
+    {
+        auto http = m_blob_client->client()->get_handle();
+        auto request = std::make_shared<flush_data_request>(filesystem, file, offset);
+        auto async_func = std::bind(&azure::storage_lite::async_executor<void>::submit, m_account, request, http, m_context);
+        return blob_client_adaptor<void>(async_func);
+    }
+
+    void adls_client::upload_file_from_stream(const std::string& filesystem, const std::string& file, std::istream& in_stream, const std::vector<std::pair<std::string, std::string>>& properties)
+    {
+        auto blob_client = m_blob_client;
+        auto async_task = [blob_client, filesystem, file, &in_stream, properties]()
+        {
+            return blob_client->upload_block_blob_from_stream(filesystem, file, in_stream, properties);
+        };
+        return blob_client_adaptor<void>(async_task);
+    }
+
+    void adls_client::download_file_to_stream(const std::string& filesystem, const std::string& file, std::ostream& out_stream)
+    {
+        return download_file_to_stream(filesystem, file, 0, 0, out_stream);
+    }
+
+    void adls_client::download_file_to_stream(const std::string& filesystem, const std::string& file, uint64_t offset, uint64_t size, std::ostream& out_stream)
+    {
+        // std::bind doesn't seem to work since download_blob_to_stream is an overloaded function.
+        auto blob_client = m_blob_client;
+        auto async_task = [blob_client, filesystem, file, offset, size, &out_stream]()
+        {
+            return blob_client->download_blob_to_stream(filesystem, file, offset, size, out_stream);
+        };
+        return blob_client_adaptor<void>(async_task);
+    }
+
+    void adls_client::delete_file(const std::string& filesystem, const std::string& file)
+    {
+        return blob_client_adaptor<void>(std::bind(&azure::storage_lite::blob_client::delete_blob, m_blob_client, filesystem, file, false));
+    }
+
+    bool adls_client::file_exists(const std::string& filesystem, const std::string& file)
+    {
+        bool blob_exists = false;
+        try
+        {
+            auto blob_properties = blob_client_adaptor<azure::storage_lite::blob_property>(std::bind(&azure::storage_lite::blob_client::get_blob_properties, m_blob_client, filesystem, file));
+            blob_exists = blob_properties.valid();
+        }
+        catch (const storage_exception& e)
+        {
+            if (e.code != 404)
+            {
+                throw;
+            }
+        }
+        if (!success() && errno == 404)
```

Comment:
>404 [](start = 35, length = 3)

Sometimes, 404 also means that credential for the authentication method is wrong. It is something worth documenting.

Reply:
`errno` doesn't keep that information. We can not tell apart the two situations, this is so bad.

Reply:
I added a comment reminding users of this.